### ISSUE TITLE
flash attention returns wrong output for non-contiguous value tensors (breaks Qwen3.5-9B E2E)

### ISSUE THREAD

**Issue by shiptux** [FlagOpen/FlagGems#5896]
## Summary

`flag_gems` flash attention (`_flash_attention_forward`, reached via `F.scaled_dot_product_attention`) produces wrong output when the **value tensor is non-contiguous** (a transpose view). The kernel appears to read V assuming a contiguous `(B, H, S, D)` layout instead of honoring its strides. ATen's flash attention handles the same input correctly.

Confirmed on current master (29e1482) and on 5.0.2, A100 (sm_80), torch 2.5.1+cu124, triton 3.6.0 (FlagTree), fp16.

## How we hit it

End-to-end validation of Qwen3.5-9B (hybrid linear/full-attention architecture) on A100: with `flag_gems.enable()`, first-step logits cosine vs stock torch dropped to **0.807** (greedy output diverges after 12 tokens). Delta-debugging over the registered ops isolated a minimal interacting pair: `{cat, _flash_attention_forward}` — each op alone is clean (cosine 0.99999), together they reproduce 0.807.

Root cause of the interaction: in the Qwen3.5 attention path (partial RoPE, `torch.cat([q_embed, q_pass], dim=-1)`), a downstream `reshape` happens to produce a **transpose view** for V when flag_gems' `cat` output feeds it (both layouts are valid; the values are identical). ATen flash handles the non-contiguous V; the flag_gems kernel does not. Captured metadata at the `_scaled_dot_product_flash_attention` boundary, first full-attention layer:

```
aten path:  value shape=(1, 4, 10, 256) stride=(10240, 2560, 256, 1) contiguous=True
gems path:  value shape=(1, 4, 10, 256) stride=(10240, 256, 1024, 1) contiguous=False
```

## Standalone repro (no model needed)

```python
import torch, torch.nn.functional as F
import flag_gems

torch.manual_seed(0)
B, Hq, Hkv, S, D = 1, 16, 4, 10, 256
q = torch.randn(B, Hq, S, D, device="cuda", dtype=torch.float16)
k = torch.randn(B, Hkv, S, D, device="cuda", dtype=torch.float16)
v = torch.randn(B, S, Hkv, D, device="cuda", dtype=torch.float16).transpose(1, 2)
# v: shape (1, 4, 10, 256), stride (10240, 256, 1024, 1) — non-contiguous

ref = F.scaled_dot_product_attention(q.float(), k.float(), v.float(),
                                     is_causal=True, enable_gqa=True).half()

def flash(q, k, v):
    with torch.nn.attention.sdpa_kernel(torch.nn.attention.SDPBackend.FLASH_ATTENTION):
        return F.scaled_dot_product_attention(q, k, v, is_causal=True, enable_gqa=True)

flag_gems.only_enable(include=["_flash_attention_forward", "_scaled_dot_product_flash_attention"])

print("non-contig V:", float((flash(q, k, v) - ref).abs().max()))            # 4.66  <-- wrong
print("contig V:    ", float((flash(q, k, v.contiguous()) - ref).abs().max())) # 0.002 <-- ok
```

Output on A100:

```
non-contig V: 4.65625
contig V:     0.001953125
```

(On 5.0.2 the include name is `flash_attention_forward`; same numbers.)

## Expected

Either honor the input strides in the kernel, or fall back to `.contiguous()` on non-contiguous q/k/v before launching (matching ATen behavior). Query/key may have the same exposure as value — worth auditing the same entry point for all three inputs.

## Environment

- FlagGems master 29e1482 (also reproduced on 5.0.2)
- NVIDIA A100-SXM4-40GB (sm_80), driver 535.161.08
- torch 2.5.1+cu124, triton 3.6.0 (FlagTree), Python 3.11, openEuler 24.03

**Comment by shiptux** [comment]
Additional evidence that this also breaks a plain GQA model (Qwen3-0.6B) with current transformers, so the impact is broader than the Qwen3.5 hybrid architecture.

## End-to-end

Qwen3-0.6B, fp16, greedy decoding, stock torch vs `flag_gems.enable()`:

| attention path | first-step logits cosine | greedy token agreement |
|---|---|---|
| default (`sdpa`) | **-0.065** (garbage output) | 0/64 |
| `attn_implementation="eager"` | 0.999996 | 64/64 |

Same numbers with transformers 5.16.0 and 5.17.0, so it is not a transformers-version effect. FlagGems 5.3.0 and 5.3.5 both reproduce.

## Operator-level isolation

Four ways of feeding the Qwen3-0.6B shapes (16 query heads, 8 KV heads, head_dim 128) into `F.scaled_dot_product_attention` under `flag_gems.enable()`, compared with stock torch:

```python
import torch, torch.nn.functional as F
torch.manual_seed(0)
B, Hq, Hkv, L, D = 1, 16, 8, 64, 128
dt = torch.float16
q = torch.randn(B, Hq, L, D, device="cuda", dtype=dt)
k = torch.randn(B, Hkv, L, D, device="cuda", dtype=dt)
v = torch.randn(B, Hkv, L, D, device="cuda", dtype=dt)
def repeat_kv(x):  # transformers' repeat_kv
    b, h, l, d = x.shape
    return x[:, :, None].expand(b, h, Hq // Hkv, l, d).reshape(b, Hq, l, d)
cases = {
    "A_gqa_flag":      lambda: F.scaled_dot_product_attention(q, k, v, is_causal=True, enable_gqa=True),
    "B_repeat_contig": lambda: F.scaled_dot_product_attention(q, repeat_kv(k).contiguous(), repeat_kv(v).contiguous(), is_causal=True),
    "C_repeat_view":   lambda: F.scaled_dot_product_attention(q, repeat_kv(k), repeat_kv(v), is_causal=True),
    "D_noncontig_v":   lambda: F.scaled_dot_product_attention(q, repeat_kv(k).contiguous(),
                           repeat_kv(v).contiguous().transpose(1, 2).contiguous().transpose(1, 2), is_causal=True),
}
ref = {n: f().float() for n, f in cases.items()}
import flag_gems; flag_gems.enable()
for n, f in cases.items():
    print(n, float((f().float() - ref[n]).abs().max()))
```

```
A_gqa_flag       0.00049   OK
B_repeat_contig  0.00049   OK
C_repeat_view    0.00049   OK
D_noncontig_v    5.98      MISMATCH
```

Only the transposed-view (non-contiguous) V case fails, which is exactly the condition described above. In transformers the projected K/V states are produced as `k_proj(x).view(B, L, H, D).transpose(1, 2)` before reaching sdpa, i.e. a transposed view is the standard layout, which is why the whole model fails under the default sdpa path while `eager` (plain matmul) is fine.

Environment: FlagGems 5.3.5 (also 5.3.0), torch 2.10.0+cu128, triton 3.6.0, transformers 5.16.0 / 5.17.0, NVIDIA A100-SXM4-40GB (sm_80), driver 535.161.08, Ubuntu 22.04, Python 3.12.

A `.contiguous()` fallback on non-contiguous q/k/v at the sdpa entry point would unblock every HF model on the default attention path until the kernel honors strides.
